zipline/result_helpers.py

```diff
@@ -0,0 +1,297 @@
+
+    def handle_message(self, message, progress_bar):
+        if 'daily_perf' not in message:
+            # Only process daily perf packets.
+            return
+
+        dt = self._last_dt = message['daily_perf']['period_close'].normalize()
+        for ext in self._extractors:
+            ext.update(dt, message)
+
+    def finalize(self):
+        merged = tz.merge(e.finalize() for e in self._extractors)
+        merged['end_date'] = self._last_dt
+        return merged
+
+
+def default_result_stream_processor():
+    return ResultStreamProcessor([
+        extract_daily_perf(),
+        extract_cumulative_perf(),
+        extract_risk(),
+        extract_orders(),
+        extract_positions(),
+        extract_transactions(),
+        extract_recorded_vars(),
+        extract_top_level_metadata(),
+    ])
+
+
+class SchematizedPayloadExtractor(object):
+
+    def __init__(self, name, path, dtypes, renames):
+        self._name = name
+        self._path = path
+        assert viewkeys(renames) <= viewkeys(dtypes), (
+            "Renames must be a subset of dtypes:\n renames=%s, dtypes=%s"
+            % (renames, dtypes)
+        )
+        self._renames = renames
+        self._dtypes = dtypes
+        self._dts = []
+        self._columns = {k: [] for k in self._dtypes}
+
+    def update(self, dt, top_level_message):
+        columns = self._columns
+        dts = self._dts
+
+        records = tz.get_in(self._path, top_level_message, no_default=True)
+        if not isinstance(records, list):
+            records = (records,)
+
+        for message in records:
+            dts.append(dt)
+            for key in columns:
+                columns[key].append(message[key])
+
+    def finalize(self):
+        index = pd.to_datetime(self._dts, utc=True)
+        dtypes = self._dtypes
+        arrays = {
+            # Apply renames here.
+            self._renames.get(k, k): _to_column(data, dtypes[k])
+            for k, data in iteritems(self._columns)
+        }
+        result = pd.DataFrame(arrays, index=index)
+        return {self._name: result}
+
+
+def _to_column(data, dtype):
+    arr = np.array(data, dtype=dtype)
+    if dtype.kind == 'M':
+        arr = pd.DatetimeIndex(arr, tz='UTC')
+    return arr
+
+
+def extract_daily_perf():
+    path = ['daily_perf']
+    dtypes = {
+        # These fields should be kept in sync with extract_cumulative_perf.
+        'capital_used':    float64_dtype,
+        'ending_cash':     float64_dtype,
+        'ending_value':    float64_dtype,
+        'pnl':             float64_dtype,
+        'portfolio_value': float64_dtype,
+        'returns':         float64_dtype,
+        # These fields are specific to daily perf.
+        'starting_value':  float64_dtype,
+        'starting_cash':   float64_dtype,
+        'period_open':     np.dtype('datetime64[ms]'),
+        'period_close':    np.dtype('datetime64[ms]'),
+        'longs_count':     int64_dtype,
+        'shorts_count':    int64_dtype,
```

Comment:
They're new relative to the Quantopian version.  Zipline actually emits a lot more fields in perf packets than we store on quantopian; most (all?) of the extra fields, like `longs_count` and `shorts_count`, can be re-derived from the source data, so I assume we don't store them downstream to save space.  I added these here because we had tests that depended on them being present in the result.